ewatercycle/analysis/__init__.py

```diff
@@ -0,0 +1,136 @@
+    fname : str, Path
+        If specified, a copy of the plot will be saved to this path.
+    dpi : int, optional
+        DPI for the plot.
+    title : str, optional
+        Title of the hydrograph.
+    discharge_units : str, optional
+        Units for the discharge data.
+    precipitation_units : str, optional
+        Units for the precipitation data.
+
+    Returns
+    -------
+    ax : :obj:`matplotlib.axes.Axes`
+        Instance of a `~matplotlib.axes.Axes` object.
+    """
+    discharge_cols = discharge.columns.drop('reference')
+    y_obs = discharge[reference]
+    y_sim = discharge[discharge_cols]
+
+    fig, (ax, ax_tbl) = plt.subplots(
+        nrows=2,
+        ncols=1,
+        dpi=dpi,
+        figsize=(10, 10),
+        gridspec_kw={'height_ratios': [3, 1]},
+    )
+
+    ax.set_title(title)
+    ax.set_ylabel(f'Discharge ({discharge_units})')
+
+    y_obs.plot(ax=ax)
+    y_sim.plot(ax=ax)
+
+    handles, labels = ax.get_legend_handles_labels()
+
+    # Add precipitation as bar plot to the top if specified
+    if precipitation is not None:
+        ax_pr = ax.twinx()
+        ax_pr.invert_yaxis()
+        ax_pr.set_ylabel(f'Precipitation ({precipitation_units})')
+
+        for pr_label, pr_timeseries in precipitation.iteritems():
+            ax_pr.bar(
+                pr_timeseries.index.values,
+                pr_timeseries.values,
+                alpha=0.4,
+                label=pr_label,
+            )
+
+        # tweak ylim to make space at bottom and top
+        ax_pr.set_ylim(ax_pr.get_ylim()[0] * (7 / 2), 0)
+        ax.set_ylim(0, ax.get_ylim()[1] * (7 / 5))
+
+        # prepend handles/labels so they appear at the top
+        handles_pr, labels_pr = ax_pr.get_legend_handles_labels()
+        handles = handles_pr + handles
+        labels = labels_pr + labels
+
+    # Put the legend outside the plot, at the bottom
+    ax.legend(handles, labels, bbox_to_anchor=(1.10, 1), loc='upper left')
+
+    # set formatting for xticks
+    date_fmt = DateFormatter("%Y-%m")
+    ax.xaxis.set_major_formatter(date_fmt)
+    ax.tick_params(axis='x', rotation=30)
+
+    # calculate metrics for data table underneath plot
+    def calc_metric(metric) -> float:
+        return y_sim.apply(metric, observed_array=y_obs)
+
+    metrs = pd.DataFrame({
+        'nse': calc_metric(metrics.nse),
+        'kge_2009': calc_metric(metrics.kge_2009),
+        'sa': calc_metric(metrics.sa),
+        'me': calc_metric(metrics.me),
+    })
+
+    # convert data in dataframe to strings
+    cell_text = [[f'{item:.2f}' for item in row[1]]
+                 for row in metrs.iterrows()]
+
+    table = ax_tbl.table(
+        cellText=cell_text,
+        rowLabels=metrs.index,
+        colLabels=metrs.columns,
+        loc='center',
+    )
+    ax_tbl.set_axis_off()
+
+    # give more vertical space in cells
+    table.scale(1, 1.5)
+
+    if fname is not None:
+        fig.savefig(fname, bbox_inches='tight', dpi=dpi)
+
+    return ax
```

Comment:
Why return only the figure axes and not the table? Consider returning the figure instance as well?

Reply:
Yeah, better to return just the `fig`, since it contains both axes 👍